Preparation for 0.6.0 (#517)
microsoft/snmalloc: docs/security/FreelistProtection.md

# Protecting meta-data

> Corrupting an allocator's meta-data is a common pattern for increasing the power of a use-after-free or out-of-bounds write vulnerabilities.  If you can corrupt the allocator's meta-data, then you can take a control gadget in one part of a system, and use it to affect other parts of the system.  There are various approaches to protecting allocator meta-data, the most common are:

Reviewer: Can you put each sentence on a separate line so that diffs are easy to read and it's easier to review please?

> * make the allocator meta-data hard to find through randomisation

Reviewer 2: Do you also want to call out separating the metadata pool from the allocated ranges? That is key both for CHERI and sandboxing, where the allocated range is considered dangerous for the allocator to interact with. It is part of the "hard to find" paradigm, but more specific and something that people don't normally think about with heap corruption attacks.

Reviewer: Agreed on putting this in the notes.  With the latest version (on 64-bit platforms), address space never moves between use as metadata and use as allocations, which means that UAFs won't arbitrarily corrupt allocator state.  This is a bigger win with CHERI (which isn't supported in 0.6, but will be in 0.7) when you don't have revocation because it means that there's no path that allows code outside of the allocator to reach metadata.  

For sandboxing use cases this is less relevant because the pagemap still has pointers to the metadata, so anything that can reach the pagemap can find the entries and we need to protect them in some other way (in the process sandboxing code, by allocating them in memory that the untrusted code doesn't have MMU mappings for, in the CHERI sandboxing case we might do this via sealing, but if the allocator is part of your TCB anyway then we may just trust it).

Author: Okay, so should I add a bullet point like

```suggestion
* make the allocator meta-data hard to find through randomisation
* use completely separate ranges of memory for meta-data and allocations
```

# Protecting meta-data

Corrupting an allocator's meta-data is a common pattern for increasing the power of a use-after-free or out-of-bounds write vulnerabilities.
If you can corrupt the allocator's meta-data, then you can take a control gadget in one part of a system, and use it to affect other parts of the system.
There are various approaches to protecting allocator meta-data, the most common are:

* make the allocator meta-data hard to find through randomisation
* surround meta-data with guard pages
* add some level of encryption/checksuming

> With the refactoring of the page table ([described earlier](./VariableSizedChunks.md)), we can put all the slab meta-data at hard to find places with randomisation and add guard pages around it.

Author: @ihaller @davidchisnall does this say enough about the meta-data protection?

```suggestion
With the refactoring of the page table ([described earlier](./VariableSizedChunks.md)), we can put all the slab meta-data in completely separate regions of memory to the allocations.
Within the meta-data regions, we add randomisation to make the data hard to find, and add large guard regions around the meta-data.
By using completely separate regions of memory for allocations and meta-data we ensure that no dangling allocation can refer to current meta-data.
This is particularly important for CHERI as it means a UAF can be used to corrupt allocator meta-data.

```

* make the allocator meta-data hard to find through randomisation
* surround meta-data with guard pages
* add some level of encryption/checksuming

With the refactoring of the page table ([described earlier](./VariableSizedChunks.md)), we can put all the slab meta-data at hard to find places with randomisation and add guard pages around it.
But there is one super important bit that still remains: free lists.

##  What are free lists?

Many allocators chain together unused allocations into a linked list.
This is super efficient as it doesn't require meta-data proportional to the number of allocations on a slab.
The disused objects can be used in either a linked stack or queue.
However, the key problem is neither randomisation or guard pages can be used to protect this data.

In snmalloc, we are using a new technique for protecting this data.

## Protecting a free queue.

The idea is remarkably simple: a doubly linked list is far harder to corrupt than a single linked list, because you can check its invariant:
```
   x.next.prev == x
```
In every kind of free list in snmalloc, we now use something a kin to this:
```

>   f(x.next).prev == g(x, f(x.next))

Reviewer 3: This looks like you are `f(x.next)` as structure pointer but below it says that `f` is XOR with random values?!

```
Here `f` is an [involution](https://en.wikipedia.org/wiki/Involution_(mathematics)) such as XORing a randomly choosen value:
```
  f(a) = a XOR k0
```
This means we don't store the `next` pointer in a way that an attacker can easily read.
The `g` function is a little more complex
```
  g(a, b) = (a XOR k1) * (b XOR k2)
```
where `k1` and `k2` are two randomly chosen 64 bit values.
This gives a value that is hard to forge that encodes the back edge relationship.

As we build the list, we add this value to the disused object, and when we consume the free list later, we check the value is correct.
Importantly, the order of construction and consumption have to be the same, which means we can only use queues, and not stacks.

> The checks give us a way to detect that the list has not been mutated.

Reviewer 3: ```suggestion
The checks give us a way to detect that the list has not been corrupted.
```
Hopefully it is mutated in useful ways...

We had to refactor the per slab free lists to be queues rather than stacks, but that is fairly straightforward.
The code for the free lists can be found here:

[Code](https://github.com/microsoft/snmalloc/blob/main/src/snmalloc/mem/freelist.h)

The idea could easily be applied to other allocators, and we're happy to discuss this.

## Finished assembly

So let's look at what costs we incur from this.
There are bits that are added to both creating the queues, and taking elements from the queues.
Here we show the assembly for taking from the queue, which is integrated into the fast path of allocation:
```x86asm
<malloc(unsigned long)>:
    lea    rax,[rdi-0x1]                       # Check for small size class
    cmp    rax,0xdfff                          # | zero is considered a large size
    ja     SLOW_SIZE                           # | to remove from fast path.
    shr    rax,0x4                             # Lookup size class in table
    lea    rcx,[size_table]                    # | 
    movzx  edx,BYTE PTR [rax+rcx*1]            # | 
    mov    rdi,rdx                             #+Caclulate index into free lists

>     shl    rdi,0x4                             #+|

Author: Actually, it is because lea can only represent shifts upto 3.

    je     SLOW_PATH_REFILL                    # |
    mov    rsi,QWORD PTR fs:0x0                # Calculate location of free list structure
    add    rsi,r8                              # | rsi = fs:[r8]
    mov    rdx,QWORD PTR fs:[r8+0x2e8]         #+Load next pointer key
    xor    rdx,QWORD PTR [rax]                 # Load next pointer
    prefetcht0 BYTE PTR [rdx]                  # Prefetch next object
    mov    QWORD PTR [rcx],rdx                 # Update head of free list
    mov    rcx,QWORD PTR [rax+0x8]             #+Check signed_prev value is correct
    cmp    rcx,QWORD PTR fs:[r8+rdi*1+0x8]     #+|
    jne    CORRUPTION_ERROR                    #+|
    lea    rcx,[rdi+rsi*1]                     #+Calculate signed_prev location
    add    rcx,0x8                             #+|  rcx = fs:[r8+rdi*1+0x8]
    mov    rsi,QWORD PTR fs:[r8+0x2d8]         #+Calculate next signed_prev value
    add    rsi,rax                             #+|
    add    rdx,QWORD PTR fs:[r8+0x2e0]         #+|
    imul   rdx,rsi                             #+|
    mov    QWORD PTR [rcx],rdx                 #+Store signed_prev for next entry.
    ret
```

> The additions specific to handling the checks are marked with `+`.

Reviewer 3: Ooops, this duplicates @nwf 's suggestion above.

As you can see the fast path is about twice the length of the fast path without protection, but only adds a single branch to the fast path, one multiplication, five additional loads, and one store.
The loads only involve one additional cache line for key material.
Overall, the cost is surprisingly low.

Note: the free list header now contains the value that `prev` should contain, which leads to slightly worse x86 codegen.
For instance the checks introduce `shl rdi,0x4`, which was previously fused with an `lea` instruction without the checks.

## Conclusion

This approach provides a strong defense against corruption of the free lists used in snmalloc.
This means all inline meta-data has corruption detection.
The check is remarkably simple for building double free detection, and has far lower memory overhead compared to using an allocation bitmap.

> [Next we show hot to randomise the layout of memory in snmalloc, and thus make it harder to guess relative address of a pair of allocations.](./Randomisation.md)

Reviewer 3: Also a duplicate.
